events/janus_gelfevh.c

```diff
@@ -0,0 +1,617 @@
+				int key = rand() % (int)(sizeof(charset) - 1);
+				randomString[n] = charset[key];
+			}
+			randomString[length] = '\0';
+		}
+	}
+	return randomString;
+}
+
+static void janus_gelfevh_connect(void) {
+	struct addrinfo *res = NULL;
+	janus_network_address addr;
+	janus_network_address_string_buffer addr_buf;
+	struct sockaddr_in servaddr;
+
+	if(getaddrinfo(backend, NULL, NULL, &res) != 0 ||
+		janus_network_address_from_sockaddr(res->ai_addr, &addr) != 0 ||
+		janus_network_address_to_string_buffer(&addr, &addr_buf) != 0) {
+		if(res)
+			freeaddrinfo(res);
+		JANUS_LOG(LOG_ERR, "Could not resolve address (%s)...\n", backend);
+		return;
+	}
+	const char *host = g_strdup(janus_network_address_string_from_buffer(&addr_buf));
+	freeaddrinfo(res);
+
+    if((sockfd = socket(AF_INET, janus_gelfevh_socket_type, 0)) < 0 ) {
+		JANUS_LOG(LOG_ERR, "Socket creation failed: %s", strerror(errno));
+		return;
+	}
+
+	memset(&servaddr, 0, sizeof(servaddr));
+
+	servaddr.sin_family = AF_INET;
+	servaddr.sin_port = htons(atoi(port));
+	servaddr.sin_addr.s_addr = inet_addr(host);
+
+	if(connect(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
+		JANUS_LOG(LOG_WARN, "Connect to Gelf host failed\n");
+		return;
+	}
+	JANUS_LOG(LOG_INFO, "Connected to GELF backend: [%s:%s]\n", host, port);
+}
+
+static int janus_gelfevh_send(char *message) {
+	if(!message) {
+		JANUS_LOG(LOG_WARN, "Message is NULL, not sending to Gelf!\n");
+		return -1;
+	}
+	/* TCP */
+	if(janus_gelfevh_socket_type == 1) {
+		unsigned int out_bytes = 0;
+		while (out_bytes < strlen(message)+1) {
+			int n = write(sockfd, message, strlen(message) + 1);
+			if (n < 0){
+				JANUS_LOG(LOG_WARN, "Unable to send message: %s\n", strerror(errno));
+				close(sockfd);
+				return -1;
+			} else if (n == 0) {
+				JANUS_LOG(LOG_WARN, "Connection has been unexpectedly closed by remote side: \n", strerror(errno));
+				close(sockfd);
+				return -2;
+			} else {
+				out_bytes += n;
+			}
+		}
+	/* UDP chunking with headers */
+	} else {
+		/* Check if we need to compress the data */
+		int len = strlen(message);
+		char *buf = message;
+		if(compress) {
+			char compressed_text[8192];
+			size_t compressed_len = 0;
+			compressed_len = janus_gzip_compress(compression,
+				message, strlen(message),
+				compressed_text, sizeof(compressed_text));
+			if(compressed_len == 0) {
+				JANUS_LOG(LOG_ERR, "Failed to compress event (%zu bytes)...\n", strlen(message));
+				/* Sending message uncompressed */
+			} else {
+				len = compressed_len;
+				buf = compressed_text;
+			}
+		}
+
+		int total = len / max_gelf_msg_len + 1;
+		if (total > MAX_GELF_CHUNKS) {
+			JANUS_LOG(LOG_WARN, "Gelf allows %d number of chunks, try increasing max_gelf_msg_len\n", MAX_GELF_CHUNKS);
+			return -1;
+		}
+		/* do we need to chunk the message */
+		if(total == 1) {
+			int n = write(sockfd, buf, len);
+			if(n < 0) {
+				JANUS_LOG(LOG_WARN, "Sending UDP message failed: %s \n", strerror(errno));
```

Comment:
This should probably be a `LOG_ERR` too. For the output, we usually print `errno` itself too, besides the description, as it makes it easier to debug calls, e.g.:

    JANUS_LOG(LOG_WARN, "Sending UDP message failed: %d (%s)\n", errno, strerror(errno));

As a side note, it looks like you're still using `write` instead of `send`, which is preferred instead, especially when errors happen.

Reply:
Please address this too, as it can be helpful when debugging issues.

Reply:
I will change it, but it is mentioned that the only difference is in the flags:

> With a zero flags argument, send() is equivalent to write(2)